lib/livebook/evaluator/io_proxy.ex

```diff
@@ -289,6 +304,42 @@ defmodule Livebook.Evaluator.IOProxy do
     end
   end
 
+  defp chars_from_input("", _, _count), do: {:eof, ""}
+
+  defp chars_from_input(input, :unicode, count) do
+    if byte_size_utf8(input) >= count do
+      chars_part(input, :unicode, count)
+    else
+      {input, ""}
+    end
+  end
+
+  defp chars_from_input(input, :latin1, count) do
+    if byte_size(input) >= count do
+      chars_part(input, :latin1, count)
+    else
+      {input, ""}
+    end
+  end
+
+  defp chars_part(chars, _, 0), do: {"", chars}
+
+  defp chars_part(<<head::utf8, tail::binary>>, :unicode, count) do
+    {chars, rest} = chars_part(tail, :unicode, count - 1)
+    {<<head::utf8>> <> chars, rest}
+  end
+
+  defp chars_part(<<head, tail::binary>>, :latin1, count) do
+    {chars, rest} = chars_part(tail, :latin1, count - 1)
+    {<<head>> <> chars, rest}
+  end
```

Comment:
You're recursively calling the function, but it's not the lass call (tail recursion), so this can lead to stack overflow. In this case we don't actually need recursion, just split at `count` bytes.

```suggestion
  defp chars_part(input, :latin1, count) do
    <<chars::binary-size(count), rest::binary>> = input
    {chars, rest}
  end
```